caffe2/utils/threadpool/pthreadpool_impl.cc

```diff
@@ -1,23 +1,23 @@
 #include "caffe2/utils/threadpool/pthreadpool.h"
-#include "caffe2/utils/threadpool/pthreadpool_impl.h"
 #include "caffe2/utils/threadpool/ThreadPool.h"
 
 
 //
 // External API
 //
 
-void pthreadpool_compute_1d(struct pthreadpool* threadpool,
+void pthreadpool_compute_1d(pthreadpool_t threadpool,
                             pthreadpool_function_1d_t function,
                             void* argument,
-                            size_t range) {
-    threadpool->pool_->run(
-      [function, argument](int threadId, size_t workId) {
-        function(argument, workId);
-      },
-      range);
+                            size_t range)
+{
+  reinterpret_cast<caffe2::ThreadPool*>(threadpool)->run(
+    [function, argument](int threadId, size_t workId) {
+      function(argument, workId);
+    },
+    range);
 }
 
-size_t pthreadpool_get_threads_count(struct pthreadpool* threadpool) {
-  return threadpool->pool_->getNumThreads();
+size_t pthreadpool_get_threads_count(pthreadpool_t threadpool) {
+  return reinterpret_cast<caffe2::ThreadPool*>(threadpool)->getNumThreads();
```

Comment:
forgive my ignorance, but how does this work?

Reply:
`reinterpret_cast` is similar to C-style cast. It is needed because I cast pointers between unrelated class hierarchies: `pthreadpool_t` is a typedef for `struct pthreadpool*`, where `struct pthreadpool` is never fully defined, and not related to `caffe2::ThreadPool`. It works because NNPACK passes threadpool pointer as it to `pthreadpool_compute...` functions, where it is cast back to `caffe2::ThreadPool*`